NIFI-12674 Modified ValidateCSV to make the schema optional if a head…

Freedom9339 opened the pull request:
…er is provided. Added validate on attribute option.

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-12674](https://issues.apache.org/jira/browse/NIFI-12674)

Made the schema optional for the ValidateCSV processor if a header is provided. In this case, only the structure of the CSV will validated, using the header to determine how many fields each line should have. 

Additionally, a new validation strategy was implemented, Validate on Attribute. This works similar to the way ValidateXML works in which the value of a given attribute of a FlowFile will be treated as the contents of a CSV file. Validation will be done on that attribute and not on the content of the FlowFile.

I would also like to note that I made the stream a variable that can be assigned to either the FlowFile content or the value of the attribute. In doing this, I removed the need for an inner method, and thus all of the variables that previously needed to be Atomic References could now be regular variables. 

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [X] Documentation formatting appears as expected in rendered files

pvillard31 commented:
Any reason for not using ValidateRecord processor if the only requirement is to confirm that the data is valid CSV without any specific constraint?

Freedom9339 commented:
@pvillard31 The ValidateRecord processor splits the input FlowFile into 2 Flowfiles, one for valid and another for invalid records. With the change to ValidateCSV, the whole file will be routed to either valid or invalid.

dan-s1 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -171,7 +174,16 @@ public class ValidateCsv extends AbstractProcessor {
+            .build();
+
+    public static final PropertyDescriptor VALIDATION_ATTRIBUTE = new PropertyDescriptor.Builder()
+            .name("validate-csv-attribute")
+            .displayName("Validation attribute")
```
```suggestion
            .name("Validation Attribute")
            .displayName("Validation Attribute")
```

dan-s1 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -458,156 +476,148 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-        final String schema = context.getProperty(SCHEMA).evaluateAttributeExpressions(flowFile).getValue();
-        final CellProcessor[] cellProcs = this.parseSchema(schema);
-        final boolean isWholeFFValidation = context.getProperty(VALIDATION_STRATEGY).getValue().equals(VALIDATE_WHOLE_FLOWFILE.getValue());
+        String schema = context.getProperty(SCHEMA).evaluateAttributeExpressions(flowFile).getValue();
```
```suggestion
        final String schema = context.getProperty(SCHEMA).evaluateAttributeExpressions(flowFile).getValue();
```

Freedom9339 replied:
The schema cannot be final because if it's null, I modify it in line 519.

dan-s1 replied:
Good point though instead of modifying it couldn't you just have
`cellProcs = this.parseSchema(newSchema.substring(0, newSchema.length() - 1));`

Freedom9339 replied:
I could, I just didn't want to make it confusing if down the line the schema variable needs to be used.

dan-s1 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -458,156 +476,148 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-        final CellProcessor[] cellProcs = this.parseSchema(schema);
-        final boolean isWholeFFValidation = context.getProperty(VALIDATION_STRATEGY).getValue().equals(VALIDATE_WHOLE_FLOWFILE.getValue());
+        String schema = context.getProperty(SCHEMA).evaluateAttributeExpressions(flowFile).getValue();
+        CellProcessor[] cellProcs = null;
+        if (schema != null) {
+            cellProcs = this.parseSchema(schema);
+        }
```
```suggestion
        final CellProcessor[] cellProcs = schema != null ? this.parseSchema(schema) : null;
```

Freedom9339 replied:
cellProcs cannot be final because I modify it in line 520 if the schema was null.

dan-s1 replied:
Why isn't that code pulled up here?

dan-s1 commented:
@exceptionfactory Can you please restart the failed job? It does not seem related to the changes. Thanks!

Freedom9339 commented:
Any updates on reviewing this change?

mattyb149 commented:
There are merge conflicts that need to be resolved

Freedom9339 commented:
@mattyb149 I've rebased against main. Thank You

mattyb149 commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -240,7 +254,11 @@ protected Collection<ValidationResult> customValidate(ValidationContext context)
+            if (schema != null) {
+                this.parseSchema(schema);
+            } else if (!headerProp.asBoolean()) {
+                throw(new Exception("Schema cannot be empty if header is false."));
```
Rather than throwing an exception, a ValidationResult should be added to a list to be returned at the end of the method.

Freedom9339 replied:
@mattyb149 Thank you for the review. A ValidationResult is created and returned once the thrown exception is catched. I could replace the throw exception with a ValidationResult, but it would be a duplicate of what the catch block already does.

https://github.com/apache/nifi/blob/9edb476aea715666a1e63d47996c911596e8eadb/nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java#L262-L270

jrsteinebrey commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -171,7 +174,16 @@ public class ValidateCsv extends AbstractProcessor {
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor VALIDATION_ATTRIBUTE = new PropertyDescriptor.Builder()
```
I would change this property name and description to be consistent with ValidateXML because I feel it will make it easier to understand how to use it. Plus it will make it easier for users who use both ValidateXML and ValidateCSV to learn one way of doing it instead of two.
This is the property in ValidateXML that to use. Just change XML to CSV.
    public static final PropertyDescriptor XML_SOURCE_ATTRIBUTE = new PropertyDescriptor.Builder()
            .name("XML Source Attribute")
            .displayName("XML Source Attribute")
            .description("The name of the attribute containing XML to be validated. If this property is blank, the FlowFile content will be validated.")
            .required(false)
            .expressionLanguageSupported(ExpressionLanguageScope.ENVIRONMENT)
            .addValidator(StandardValidators.ATTRIBUTE_KEY_VALIDATOR)
            .build();

Freedom9339 replied:
I've made this change.

jrsteinebrey commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -171,7 +174,16 @@ public class ValidateCsv extends AbstractProcessor {
             .required(true)
             .defaultValue(VALIDATE_WHOLE_FLOWFILE.getValue())
-            .allowableValues(VALIDATE_LINES_INDIVIDUALLY, VALIDATE_WHOLE_FLOWFILE)
+            .allowableValues(VALIDATE_LINES_INDIVIDUALLY, VALIDATE_WHOLE_FLOWFILE, VALIDATE_ATTRIBUTE_AS_CSV)
```
Remove VALIDATE_ATTRIBUTE_AS_CSV from VALIDATION_STRATEGY and use the CSV_SOURCE_ATTRIBUTE property described below to trigger validating an attribute. Using CSV_SOURCE_ATTRIBUTE removes the need for VALIDATE_ATTRIBUTE_AS_CSV.  VALIDATE_ATTRIBUTE_AS_CSV does not seem to logically belong with the other two existing options in VALIDATION_STRATEGY.
Sidenote: Can the Attribute to Validate as CSV contain more than 1 line of csv data?

Freedom9339 replied:
I've made this change. Also, yes, the attribute can contain as many lines as the attribute value can hold.

jrsteinebrey commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -458,155 +473,147 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+
+        InputStream stream;
+        if (context.getProperty(CSV_SOURCE_ATTRIBUTE).isSet()) {
+            stream = new ByteArrayInputStream(flowFile.getAttribute(context.getProperty(CSV_SOURCE_ATTRIBUTE).getValue()).getBytes(StandardCharsets.UTF_8));
```
@Freedom9339  Thanks for making that change. It turned out well except you need to call .evaluateAttributeExpressions() after the .getProperty(CSV_SOURCE_ATTRIBUTE) call. After that, the change looks complete.

Freedom9339 replied:
Done.

jrsteinebrey replied:
@Freedom9339 Thanks for your contribution.

mattyb149 commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -120,11 +120,20 @@ public class ValidateCsv extends AbstractProcessor {
                     + "processors to apply. The following cell processors are allowed in the schema definition: "
                     + allowedOperators.toString() + ". Note: cell processors cannot be nested except with Optional.")
-            .required(true)
+            .required(false)
```
This should have a `dependsOn(HEADER, "false")`, unless you can provide the explicit schema yet there be a header line that should be ignored.

mattyb149 replied:
I think the "unless" part is true here. But now there are merge conflicts, sorry I lost track of this

Freedom9339 replied:
I've updated the documentation.

exceptionfactory commented:
@mosermw noting that you removed the `stale` label, are you planning to review this pull request? Although this has had some activity, the lack involvement over four months seems to indicate that it should be closed, with the option to revisit if there is a positive comment on the Jira issue that a committer is planning to review and guide this to completion.

mosermw commented:
> @mosermw noting that you removed the `stale` label, are you planning to review this pull request?

It looks like Matt B. made some comments that should be addressed, along with the merge conflicts.  I contacted the PR author to ask if they still want this change, and will be happy to work with them to finish this.

exceptionfactory commented:
Thanks @mosermw!

Freedom9339 commented:
@mosermw @exceptionfactory @mattyb149 Sorry about loosing track of this. I've rebased and addressed the comments provided.

mosermw commented:
Reviewing ...

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -449,155 +465,147 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+        FlowFile invalidFF = null;
+        FlowFile validFF = null;
+        String validationError = null;
+        final AtomicReference<Boolean> isFirstLineValid = new AtomicReference<Boolean>(true);
+        final AtomicReference<Boolean> isFirstLineInvalid = new AtomicReference<Boolean>(true);
```
A recent commit used diamond notation for the constructor and I suspect your rebase didn't pick this up.  This is especially important since new PMD rules were recently committed that checks for this.
```suggestion
        final AtomicReference<Boolean> isFirstLineValid = new AtomicReference<>(true);
        final AtomicReference<Boolean> isFirstLineInvalid = new AtomicReference<>(true);
```

Freedom9339 replied:
Done.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestValidateCsv.java:
Please add a unit test where the CSV Source Attribute is defined but is missing from the flowfile.  Also verify that flowfile content is not modified when validating an attribute.

Freedom9339 replied:
Done.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestValidateCsv.java:
Please add a unit test to verify flowfile goes to "valid" when validating an attribute and the attribute value is missing or empty and a schema is defined.  This explicitly defines expected behavior when there is no CSV data to validate.  Also verify that flowfile content is not modified when validating an attribute.

Freedom9339 replied:
Done.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestValidateCsv.java:
Please add a unit test to verify a 0 byte flowfile goes to "valid" when validating flowfile content and a schema is defined.  This explicitly defines expected behavior when there is no CSV data to validate.  This doesn't test any changes you made in this PR but would really help fill out this unit test.

Freedom9339 replied:
Done.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -449,155 +465,147 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+
+        InputStream stream;
+        if (context.getProperty(CSV_SOURCE_ATTRIBUTE).isSet()) {
+            stream = new ByteArrayInputStream(flowFile.getAttribute(context.getProperty(CSV_SOURCE_ATTRIBUTE).evaluateAttributeExpressions().getValue()).getBytes(StandardCharsets.UTF_8));
```
The call to flowFile.getAttribute() can return null if the attribute does not exist on the flowfile.  If that happens, this line will cause a NullPointerException.

Freedom9339 replied:
Fixed.

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -204,6 +214,7 @@ public class ValidateCsv extends AbstractProcessor {
             .description("FlowFiles that are not valid according to the specified schema are routed to this relationship")
             .build();
 
+
```
This extra blank line doesn't appear to be necessary.

Freedom9339 replied:
Removed.

Freedom9339 commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
Done.

Freedom9339 commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -98,6 +97,7 @@ public class ValidateCsv extends AbstractProcessor {
             "Strlen", "StrMinMax", "StrNotNullOrEmpty", "StrRegEx", "Unique", "UniqueHashCode", "IsIncludedIn"
     );
 
+
```
Removed

Freedom9339 commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateCsv.java:
```diff
@@ -231,7 +243,11 @@ protected Collection<ValidationResult> customValidate(ValidationContext context)
+            if (schema != null) {
+                this.parseSchema(schema);
+            } else if (!headerProp.asBoolean()) {
+                throw(new Exception("Schema cannot be empty if header is false."));
```
Done.

mosermw commented:
Thanks for making these changes @Freedom9339.  One or two last things that I noticed.

When validating a CSV Source Attribute, we can set the "Validation Strategy = Line by line".  When we do this, the flowfile content is replaced with the value of the CSV attribute.  This is clearly not good!  I cannot think of a use case where we want to allow "Validation Strategy = Line by line" on an attribute ... it should always validate the whole attribute and not change anything.

We can make the CSV Source Attribute depend on "Validation Strategy = FlowFile" but then we would have to defend ourselves from the case when someone enters a CSV Source Attribute and then changes the Validation Strategy.  The CSV Source Attribute will disappear and I'm worried that it will still be set and we will still try to use it.

So, if using dependsOn() gets too complicated, just add something to customValidate to disallow CSV Source Attribute with "Validation Strategy = Line by line"

mosermw commented on nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestValidateCsv.java:
```diff
@@ -164,6 +168,107 @@ public void testBigDecimalBoolCharIntLong() {
+        runner.setProperty(ValidateCsv.HEADER, "true");
+
+        runner.enqueue("bigdecimal,bool,char,integer,long\r\n10.0001,true,c,1,92147483647");
+        runner.enqueue(new byte[0]);
```
This second runner.enqueue() isn't necessary.

Freedom9339 replied:
Removed.

Freedom9339 commented:
@mosermw I've added the depends on and made the validate attribute only show when using whole ff validation. I've added a check to make sure whole ff validation is being used before using the attribute as a source.

mosermw commented:
Thanks for sticking with this @Freedom9339.  Everything looks goo now.

+1 merging.
